_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -96,4 +96,188 @@ __P.S.__ I'm also open to your comments and suggestions. Especially if you have
+
+```cpp
+for(r=0;r<rows;r++){
+  for(c=0;c<cols;c++){
+      dot[c] = 0.0;
+      sum[c] = 0.0;
+      for(cc=(-2);cc<=2;cc++){
+        if(((c+cc) >= 0) && ((c+cc) < cols)){
+            dot[c] += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+            sum[c] += kernel[center+cc];
+        }
+      }
+      tempim[r*cols+c] = dot[c]/sum[c];
+  }
+}
+```
+
+We converted temporary values `dot` and `sum` into arrays. By doing this, their values are preserved accross loop iterations. This allows us to break the loop over `c` into two loops:
+
+```cpp
+for(r=0;r<rows;r++){
+  for(c=0;c<cols;c++){
+      dot[c] = 0.0;
+      sum[c] = 0.0;
+  }
+  for(c=0;c<cols;c++){
+      for(cc=(-2);cc<=2;cc++){
+        if(((c+cc) >= 0) && ((c+cc) < cols)){
+            dot[c] += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+            sum[c] += kernel[center+cc];
+        }
+      }
+  }
+  for(c=0;c<cols;c++){
+      tempim[r*cols+c] = dot[c]/sum[c];
+  }
+}
+```
+
+Loop over `c` and loop over `cc` are now perfectly nested and they can be interchanged.
+
+The second loop nest looks similar to the first one:
+
+```cpp
+for(c=0;c<cols;c++){
+  for(r=0;r<rows;r++){
+      sum = 0.0;
+      dot = 0.0;
+      for(rr=(-center);rr<=center;rr++){
+        if(((r+rr) >= 0) && ((r+rr) < rows)){
+            dot += tempim[(r+rr)*cols+c] * kernel[center+rr];
+            sum += kernel[center+rr];
+        }
+      }
+      (*smoothedim)[r*cols+c] = (short int)(dot*BOOSTBLURFACTOR/sum + 0.5);
+  }
+}
+```
+
+The access pattern in the innermost loop is bad this time. Access to the array `tempim` is with a stride `cols`. Everytime the value of `rr` increases by one we are accessing the elements which is `cols` places away from the previous element. This access pattern is bad from the performance point of view. To get rid of it, we would need to do a complex loop interchange to move the loop over `c` to the innermost position.
+
+To do this, we first interchange the loop over `c` and the loop over `r`. Then we interchange the loop over `c` and loop over `rr`. The final solution looks like this:
+
+```cpp
+for (r = 0; r<rows; r++){
+  for(c=0;c<cols;c++){
+      dot[c] = 0.0;
+      sum[c] = 0.0;
+  }
+  for (rr = (-center); rr <= center; rr++){
+    if (((r + rr) >= 0) ＆＆ ((r + rr) < rows)){
+      for (c = 0; c<cols; c++){
+        dot[c] += tempim[(r + rr) * cols + c] * kernel[center + rr];
+        sum[c] += kernel[center + rr];
+      }
+    }
+  }
+  for (c = 0; c < cols; c++) {
+    (*smoothedim)[r*cols + c] = (short int)(dot[c]*BOOSTBLURFACTOR / sum[c] + 0.5);
+  }
+}
+```
+
+One of the contestants also noted that the `sum[c]` has a fixed value `1` for all the pixels except the pixels at the edge of the image. This would allow us to get rid of the expensive division when `sum[c]` is `1` in the calculation of values for `smoothedim`.
```

Comment:
Links to the code would be great. I will ask all the guys to start pushing their code into the perf_challenge4 repo.

Reply:
Can you update this when all the code changes have landed.